Fix #114 - orphaned colors

jmarrec opened the pull request:
Fix #114 - orphaned colors

`RenderingColor` was a `ResourceObject` before.

Now it is a child, achieved by making it a `ModelObject` and having the Parent as a `ParentObject` (or one of its child classes).

Wrote Gtests too.

Review assignee: @macumber please

macumber commented on openstudiocore/src/model/BuildingUnit.cpp:
```diff
@@ -80,6 +80,20 @@ namespace detail {
 
+  std::vector<ModelObject> BuildingUnit_Impl::children() const {
+    std::vector<ModelObject> result;
+    if( boost::optional<RenderingColor> r = this->renderingColor() ) {
```
@jmarrec on this one (and story, etc) can we be explicit and add a comment that we want this to only parent the colors?  i dont want someone to think we just forgot to parent the spaces and add that in later.

jmarrec replied:
Not 100% sure I follow here, BuildingUnit doesn't have any children aside from the rendering color, and I think the other that do have a rendering color already define their children properly.

macumber commented on openstudiocore/src/model/RenderingColor_Impl.hpp:
```diff
@@ -40,13 +40,8 @@ class RenderingColor;
-
-
+  /** RenderingColor_Impl is a ModelObject_Impl that is the implementation class for RenderingColor.*/
+  class MODEL_API RenderingColor_Impl : public ModelObject_Impl {
```
@jmarrec RenderingColor should implement the parent call here.

jmarrec replied:
Ok, I'll override ModelObject::parent() and implement it then. thanks for the catch

macumber commented on openstudiocore/src/model/ThermalZone.cpp:
```diff
@@ -2361,6 +2366,11 @@ namespace detail {
     }
 
+    if( auto t_color = renderingColor() ) {
+      auto colorClone = t_color->clone(model).cast<RenderingColor>();
```
@jmarrec all objects need to do this in clone right?

jmarrec replied:
I guess I should (or at least could) make rendering color a child of ThermalZone. ThermalZone is an HVACComponent, but that's also a ParentObject. I must have got thrown off by that.

macumber commented on openstudiocore/src/model/test/ThermalZone_GTest.cpp:
```diff
@@ -809,3 +810,32 @@ TEST_F(ModelFixture, ThermalZone_ZoneControlHumidistat)
+  // Clone
+  ThermalZone thermalZone2(m);
+  RenderingColor renderingColor2(m);
+  EXPECT_TRUE(thermalZone2.setRenderingColor(renderingColor2));
```
@jmarrec setRenderingColor should make sure that each color has 0 or 1 parent.  It should clone as needed?  This should also be done at version translation

jmarrec replied:
I'll implement that in all classes that have rendering colors in setRenderingColor.
It'll make it check if it already has a parent, in which case Warn and return false as discussed.

macumber commented:
I think this PR has some more issues to be addressed.  If the color is going to be a child then it needs to respect the rules of 0-1 parent only.

jmarrec commented:
@macumber, I think you can do another pass at reviewing this. Changelog:

* I implemented parent and setParent (to automatically connect the cloned color to the cloned parent) in RenderingColor
* I enforce the 0 or 1 parent max rule in each of the appropriate ParentObjects setRenderingColor
* I added VT to create a new color if a color is shared by more than one object

jmarrec commented:
Here's a ruby test for the VT stuff:

run this with 2.6.1:

```ruby
require 'openstudio'
include OpenStudio::Model

m = Model.new

b1 = BuildingUnit.new(m)
b2 = BuildingUnit.new(m)
r_b = RenderingColor.new(m)
b1.setRenderingColor(r_b)
b2.setRenderingColor(r_b)


b1 = BuildingStory.new(m)
b2 = BuildingStory.new(m)
r_b = RenderingColor.new(m)
b1.setRenderingColor(r_b)
b2.setRenderingColor(r_b)


b1 = Construction.new(m)
b2 = Construction.new(m)
r_b = RenderingColor.new(m)
b1.setRenderingColor(r_b)
b2.setRenderingColor(r_b)

b1 = CFactorUndergroundWallConstruction.new(m)
b2 = CFactorUndergroundWallConstruction.new(m)
r_b = RenderingColor.new(m)
b1.setRenderingColor(r_b)
b2.setRenderingColor(r_b)

b1 = FFactorGroundFloorConstruction.new(m)
b2 = FFactorGroundFloorConstruction.new(m)
r_b = RenderingColor.new(m)
b1.setRenderingColor(r_b)
b2.setRenderingColor(r_b)

b1 = ConstructionWithInternalSource.new(m)
b2 = ConstructionWithInternalSource.new(m)
r_b = RenderingColor.new(m)
b1.setRenderingColor(r_b)
b2.setRenderingColor(r_b)

b1 = WindowDataFile.new(m)
b2 = WindowDataFile.new(m)
r_b = RenderingColor.new(m)
b1.setRenderingColor(r_b)
b2.setRenderingColor(r_b)



b1 = SpaceType.new(m)
b2 = SpaceType.new(m)
r_b = RenderingColor.new(m)
b1.setRenderingColor(r_b)
b2.setRenderingColor(r_b)


b1 = ThermalZone.new(m)
b2 = ThermalZone.new(m)
r_b = RenderingColor.new(m)
b1.setRenderingColor(r_b)
b2.setRenderingColor(r_b)

n_objs = m.getBuildingUnits.size + m.getBuildingStorys.size + m.getConstructions.size + m.getCFactorUndergroundWallConstructions.size + m.getFFactorGroundFloorConstructions.size + m.getConstructionWithInternalSources.size + m.getWindowDataFiles.size + m.getSpaceTypes.size + m.getThermalZones.size

n_colors = m.getRenderingColors.size

puts "Created #{n_objs} Objects and #{n_colors} RenderingColors"



m.save('example261.osm', true)
```

=> Created 18 Objects and 9 RenderingColors


Then run this with your local build:

```ruby
require '/home/julien/Software/Others/OS-build/Products/ruby/openstudio.so'

def osload(path)
  translator = OpenStudio::OSVersion::VersionTranslator.new
  ospath = OpenStudio::Path.new(path)
  model = translator.loadModel(ospath)
  if model.empty?
      raise "Path '#{path}' is not a valid path to an OpenStudio Model"
  else
      model = model.get
  end
  return model
end


m = osload('example261.osm')

n_objs = m.getBuildingUnits.size + m.getBuildingStorys.size + m.getConstructions.size + m.getCFactorUndergroundWallConstructions.size + m.getFFactorGroundFloorConstructions.size + m.getConstructionWithInternalSources.size + m.getWindowDataFiles.size + m.getSpaceTypes.size + m.getThermalZones.size

n_colors = m.getRenderingColors.size

puts "Ended with #{n_objs} Objects and #{n_colors} RenderingColors"

m.save('VT.osm', true)
```

=> Ended with 18 Objects and 18 RenderingColors

macumber commented on openstudiocore/src/model/ThermalZone.cpp:
```diff
@@ -2376,6 +2386,13 @@ namespace detail {
 
+    // Note: Even though rendering color is a children, ParentObject::clone isn't called by HVACComponent::clone
+    // So we need to explicitly clone the rendering color
+    if( auto t_color = renderingColor() ) {
```
@jmarrec is ThermalZone the only color-able object that has a custom clone?

jmarrec replied:
Yes, because `ThermalZone::clone` is the only one that never calls `ParentObject_Impl::clone`

macumber commented on openstudiocore/src/osversion/VersionTranslator.cpp:
```diff
@@ -4099,6 +4099,51 @@ std::string VersionTranslator::update_2_6_1_to_2_6_2(const IdfFile& idf_2_6_1, c
 
+  // We want to enforce that each rendering color has only 0 or 1 parent
+  // Use a set to check if we have already encountered this rendering color
+  std::set<std::string> renderingColorHandlesSet;
```
@jmarrec with an IDD change this version translation is required but also more straightforward

jmarrec commented:
@macumber what IDD change are you referring to?

macumber commented:
@jmarrec Remove `Group Rendering Name` or `Surface Rendering Name` from `OS:BuildingStory`, `OS:BuildingUnit`, `OS:SpaceType`, `OS:Construction`, `OS:Construction:CfactorUndergroundWall`, `OS:Construction:FfactorGroundFloor`, `OS:Construction:InternalSource`, `OS:Construction:WindowDataFile`, `OS:LightingSimulationZone`, `OS:ThermalZone`.

Add this field to `OS:Rendering:Color`:
```
  A2,  \field Parent Name
       \type object-list
       \required-field
       \object-list BuildingStoryNames
       \object-list BuildingUnitNames
       \object-list SpaceTypeNames
       \object-list ConstructionNames
       \object-list LightingSimulationZoneNames
       \object-list ThermalZoneNames
       \object-list WeekScheduleNames ! I think this is unused and could be dropped
```

As a side note, I think the `ScheduleRenderingNames` field in `OS:Schedule:Day` is unused.
